src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp

```diff
@@ -0,0 +1,129 @@
+{
+    ChipLogProgress(AppServer, "UserDirectedCommissioningServer::OnMessageReceived");
+
+    PacketHeader packetHeader;
+
+    ReturnOnFailure(packetHeader.DecodeAndConsume(msg));
+
+    if (packetHeader.GetFlags().Has(Header::FlagValues::kEncryptedMessage))
+    {
+        ChipLogError(AppServer, "UDC encryption flag set - ignoring");
+        return;
+    }
+
+    PayloadHeader payloadHeader;
+    ReturnOnFailure(payloadHeader.DecodeAndConsume(msg));
+
+    char instanceName[chip::Mdns::kMaxInstanceNameSize + 1];
+    size_t instanceNameLength =
+        (msg->DataLength() > (chip::Mdns::kMaxInstanceNameSize)) ? chip::Mdns::kMaxInstanceNameSize : msg->DataLength();
+    msg->Read(Uint8::from_char(instanceName), instanceNameLength);
+
+    instanceName[instanceNameLength] = '\0';
+
+    ChipLogProgress(AppServer, "UDC instance=%s", instanceName);
+
+    UDCClientState * client = mUdcClients.FindUDCClientState(instanceName);
+    if (client == nullptr)
+    {
+        ChipLogProgress(AppServer, "UDC new instance state received");
+
+        CHIP_ERROR err;
+        err = mUdcClients.CreateNewUDCClientState(instanceName, &client);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(AppServer, "UDC error creating new connection state");
+            return;
+        }
+
+        // Call the registered InstanceNameResolver, if any.
+        if (mInstanceNameResolver != nullptr)
+        {
+            mInstanceNameResolver->FindCommissionableNode(instanceName);
+        }
+        else
+        {
+            ChipLogError(AppServer, "UserDirectedCommissioningServer::OnMessageReceived no mInstanceNameResolver registered");
+        }
+    }
+
+    mUdcClients.MarkUDCClientActive(client);
+}
+
+void UserDirectedCommissioningServer::SetUDCClientProcessingState(char * instanceName, UDCClientProcessingState state)
+{
+    UDCClientState * client = mUdcClients.FindUDCClientState(instanceName);
+    if (client == nullptr)
+    {
+        // printf("SetUDCClientProcessingState new instance state received\n");
+        CHIP_ERROR err;
+        err = mUdcClients.CreateNewUDCClientState(instanceName, &client);
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(AppServer,
+                         "UserDirectedCommissioningServer::SetUDCClientProcessingState error creating new connection state");
+            return;
+        }
+    }
+
+    ChipLogDetail(AppServer, "SetUDCClientProcessingState instance=%s new state=%d", instanceName, (int) state);
+
+    client->SetUDCClientProcessingState(state);
+
+    mUdcClients.MarkUDCClientActive(client);
+
+    return;
+}
+
+void UserDirectedCommissioningServer::OnCommissionableNodeFound(const Mdns::DiscoveredNodeData & nodeData)
+{
+    UDCClientState * client = mUdcClients.FindUDCClientState(nodeData.instanceName);
+    if (client != nullptr && client->GetUDCClientProcessingState() == UDCClientProcessingState::kDiscoveringNode)
```

Comment:
Because this executes for all browse callbacks, I don't think it's sufficient to only check `GetUDCClientProcessingState`.  I think this also needs to filter on the `nodeData` service type and only act when `_matterc._udp` services are encountered.

Without this, it appears that if the controller initiates some other browse activity immediately prior to the UDC request, `OnUserDirectedCommissioningRequest` could be passed other unrelated services.

Reply:
This callback is only invoked today by DeviceCommissioner (which is _matterc), a check is done to ensure instanceName matches the one we are looking for (one that is in the UDC state cache with state of "kDiscoveringNode") so I think we are fine there. 

I'm open to adding more checks if needed based upon testing with real devices but I'd prefer to wait until then.